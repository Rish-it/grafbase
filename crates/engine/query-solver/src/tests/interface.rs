use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/simple-interface-object/a")
    B @join__graph(name: "b", url: "http://localhost:4200/simple-interface-object/b")
    C @join__graph(name: "c", url: "http://localhost:4200/simple-interface-object/c")
}

type Admin implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    isActive: Boolean!
    name: String!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
    @join__type(graph: C)
{
    users: [NodeWithName!]! @join__field(graph: A)
    anotherUsers: [NodeWithName] @join__field(graph: B)
    accounts: [Account] @join__field(graph: B)
}

type Regular implements Account
    @join__implements(graph: A, interface: "Account")
    @join__type(graph: A, key: "id")
{
    id: ID!
    isMain: Boolean!
    name: String!
    isActive: Boolean!
}

type User implements NodeWithName
    @join__implements(graph: A, interface: "NodeWithName")
    @join__type(graph: A, key: "id")
{
    id: ID!
    name: String
    age: Int
    username: String
}

interface Account
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
    @join__type(graph: C, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String! @join__field(graph: B)
    isActive: Boolean! @join__field(graph: C)
}

interface NodeWithName
    @join__type(graph: A, key: "id")
    @join__type(graph: B, key: "id", isInterfaceObject: true)
{
    id: ID!
    name: String @join__field(graph: A)
    username: String @join__field(graph: B)
}
"#;

#[test]
fn interface_field_providing_object_field() {
    // age is coming from subgraph A, but needs User.id for this. `anotherUsers` returns an
    // interface though, so need to retrieve the `NodeWithName.id` as an alternative for `User.id`
    assert_solving_snapshots!(
        "interface_field_providing_object_field",
        SCHEMA,
        r#"
        query {
          anotherUsers {
            ... on User {
              age
            }
          }
        }
        "#
    );
}
