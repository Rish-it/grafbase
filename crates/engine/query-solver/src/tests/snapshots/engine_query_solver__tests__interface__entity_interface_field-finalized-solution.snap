---
source: crates/engine/query-solver/src/tests/interface.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#a\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"users\" ]\n    3 [ label = \"age\" ]\n    4 [ label = \"FedEntity#b\", color=royalblue,shape=parallelogram ]\n    5 [ label = \"username\" ]\n    6 [ label = \"*id*\" ]\n    0 -> 1 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    4 -> 6 [ label = \"\", color=orangered,arrowhead=inv ]\n    2 -> 3 [ label = \"\" ]\n    2 -> 4 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    1 -> 2 [ label = \"\" ]\n    2 -> 6 [ label = \"\" ]\n    4 -> 5 [ label = \"\" ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#a" ]
    2 [ label = "users" ]
    3 [ label = "age" ]
    4 [ label = "FedEntity#b" ]
    5 [ label = "username" ]
    6 [ label = "*id*" ]
    0 -> 1 [ label = "QueryPartition" ]
    4 -> 6 [ label = "RequiredBySubgraph" ]
    2 -> 3 [ label = "Field" ]
    2 -> 4 [ label = "QueryPartition" ]
    1 -> 2 [ label = "Field" ]
    2 -> 6 [ label = "Field" ]
    4 -> 5 [ label = "Field" ]
}
